libcontainer/nsenter/nsexec.c

```diff
@@ -1,471 +1,750 @@
-		exit(1);
-	}
+	/* TODO: Currently we aren't dealing with child deaths properly. */
+
+	switch (setjmp(env)) {
+	/*
+	 * Stage 0: We're in the parent. Our job is just to create a new child
+	 *          (stage 1: JUMP_CHILD) process and write its uid_map and
+	 *          gid_map. That process will go on to create a new process, then
+	 *          it will send us its PID which we will send to the bootstrap
+	 *          process.
+	 */
+	case JUMP_PARENT: {
+			int len;
+			pid_t child;
+			char buf[JSON_MAX];
 
-	// Retrieve data
-	int  nl_total_size = NLMSG_PAYLOAD(&nl_msg_hdr, 0);
-	char data[nl_total_size];
+			/* For debugging. */
+			prctl(PR_SET_NAME, (unsigned long) "runc:[0:PARENT]", 0, 0, 0);
 
-	if ((len = read(pipenum, data, nl_total_size)) != nl_total_size) {
-		pr_perror("Failed to read netlink payload, %d != %d", len,
-			  nl_total_size);
-		exit(1);
-	}
+			/* Start the process of getting a container. */
+			child = clone_parent(&env, JUMP_CHILD);
+			if (child < 0)
+				bail("unable to fork: child_func");
 
-	jmp_buf	env;
-	int	syncpipe[2] = {-1, -1};
-	struct	nsenter_config config = process_nl_attributes(pipenum,
-						data, nl_total_size);
+			/* State machine for synchronisation with the children. */
+			while (true) {
+				enum sync_t s;
 
-	// required clone_flags to be passed
-	if (config.cloneflags == -1) {
-		pr_perror("Missing clone_flags");
-		exit(1);
-	}
-	// prepare sync pipe between parent and child. We need this to let the
-	// child know that the parent has finished setting up
-	if (pipe(syncpipe) != 0) {
-		pr_perror("Failed to setup sync pipe between parent and child");
-		exit(1);
-	}
+				/* This doesn't need to be global, we're in the parent. */
+				int syncfd = syncpipe[1];
 
-	if (setjmp(env) == 1) {
-		// Child
-		uint8_t s = 0;
-		int	consolefd = config.consolefd;
+				if (read(syncfd, &s, sizeof(s)) != sizeof(s))
+					bail("failed to sync with child: next state");
 
-		// close the writing side of pipe
-		close(syncpipe[1]);
+				switch (s) {
+				case SYNC_ERR: {
+						/* We have to mirror the error code of the child. */
+						int ret;
 
-		// sync with parent
-		if ((read(syncpipe[0], &s, 1) != 1) || (s != 1)) {
-			pr_perror("Failed to read sync byte from parent");
-			exit(1);
-		}
+						if (read(syncfd, &ret, sizeof(ret)) != sizeof(ret))
+							bail("failed to sync with child: read(error code)");
 
-		if (setsid() == -1) {
-			pr_perror("setsid failed");
-			exit(1);
-		}
+						exit(ret);
+					}
+					break;
+				case SYNC_USERMAP_PLS:
+					/* Enable setgroups(2) if we've been asked to. */
+					if (config.is_setgroup)
+						update_setgroups(child, SETGROUPS_ALLOW);
+
+					/* Set up mappings. */
+					update_uidmap(child, config.uidmap, config.uidmap_len);
+					update_gidmap(child, config.gidmap, config.gidmap_len);
+
+					s = SYNC_USERMAP_ACK;
+					if (write(syncfd, &s, sizeof(s)) != sizeof(s)) {
+						kill(child, SIGKILL);
```

Comment:
What do you mean? We didn't `waitpid` before, and I don't think we can because the processes are cloned with `CLONE_PARENT | SIGCHLD`.